src/jsMain/kotlin/net/kyori/adventure/webui/js/StyleButtons.kt

```diff
@@ -0,0 +1,107 @@
+        // This represents the case where the user's selection contains text which is immediately surrounded by the
+        // relevant tags, so we remove those tags and leave the text selected as before.
+        // Such as: text <b>|bold|</b> more text        (where | marks the edges of a selection)
+        // This is very important for being able to "toggle off" a style right after selecting one.
+        // We must "widen" the selection first so that the tags get replaced too with insertText
+        inputBox.insertText(selection, tagPrefix.length, tagSuffix.length)
+        newSelStart -= tagPrefix.length
+        newSelEnd -= tagPrefix.length
+    } else if (selection.startsWith(tagPrefix) && selection.endsWith(tagSuffix)) {
+        // This represents the case where the user's selection perfectly encapsulates the relevant opening and closing
+        // tag, so we remove that tag and leave only the text which was between said tags
+        // Such as: text |<b>bold</b>| more text        (where | marks the edges of a selection)
+        inputBox.insertText(selection.removeSurrounding(tagPrefix, tagSuffix))
+        newSelEnd -= tagPrefix.length + tagSuffix.length
+    } else {
+        // If there don't seem to be already existing relevant tags near the selection, we add them!
+        if (tag.beforeCursor.isBlank()) {
+            // This is a simple surrounding tag such as <b></b>. We make sure to keep exactly the same area selected as
+            // was selected before hitting the style button.
+            inputBox.insertText(tagPrefix + selection + tagSuffix)
+            newSelStart += tagPrefix.length
+            newSelEnd += tagPrefix.length
+        } else {
+            // This is a tag which requires more user input, such as <input:open_url...>. For example, in the open_url
+            // case, we helpfully position the cursor right where the url would go, so it can be typed right away.
+            inputBox.insertText(tag.beforeCursor + tag.placeholder + tag.beforeSel + selection + tagSuffix)
+            newSelStart += tag.beforeCursor.length
+            newSelEnd = newSelStart + tag.placeholder.length
+        }
+    }
+
+    // This makes sure the websocket stuff is actually fired so the preview stays in sync.
+    inputBox.dispatchEvent(CustomEvent("change", CustomEventInit(bubbles = true, cancelable = true)))
+    inputBox.focus()
+    inputBox.setSelectionRange(newSelStart, newSelEnd)
+}
+
+/* Apparently execCommand is deprecated and non-standard and so is "insertText", however, it seems to be the only
+ * real way to manipulate the text box content while also allowing for undoing without hijacking all the key events
+ * or something. Being nonstandard, we should *probably* have a fallback which just sets the content (and thereby
+ * nuking the undo-buffer, but oh well)
+ * Also, not supported by Internet Explorer, but like, if you're using IE that's a 'you' problem.
```

Comment:
Nobody cares about IE don't worry. Half the features I use only work on latest chrome/Firefox anyway so we clearly don't care about compatibility 😬